### PR TITLE
[grid][java] browser containers provisioned in dynamic grid can get hostconfig from node-docker

### LINKED ISSUE #1642: [🚀 Feature]: Allows  to configure volume mappings for dynamically generated containers of Dynamic Grid
### Feature and motivation

I want volume mapping to upload files and deliver camera video streams

### Usage example

```yml
[docker]

configs = [
    "selenium/standalone-firefox:4.3.0-20220726", "{\"browserName\": \"firefox\",\"volumes\":[\"/data:/data\"]}",
]
```

```python
options.add_argument('--allow-file-access-from-files')
options.add_argument('--use-fake-ui-for-media-stream')
options.add_argument('--use-fake-device-for-media-stream')
options.add_argument('--use-file-for-fake-video-capture=/data/test.y4m')
options.add_argument('--use-file-for-fake-audio-capture=/data/test.wav')
```

### LINKED ISSUE #1812: [🚀 Feature]: Ability to update hosts file on browser nodes in dynamic selenium grid
### Feature and motivation

I would like the ability to add hosts file entry to the browser nodes that get launched in the dynamic grid.

It would be good if below config is supported on the docker-compose for node-docker section and the same can be used to launch new browser nodes.

extra_hosts:
  - "cust.examplecompany.com:10.8.23.76"

### Usage example

We use selenium grid to test an internally hosted application which can be accessed using a custom dns entry in the hosts file where the browser is getting launched.

However, with dynamic selenium grid - I don't see an option to add this host file entry to the browser node.

### LINKED ISSUE #2098: [🚀 Feature]: Allow child node containers provisioned in dynamic grid to copy resolv.conf from node-docker container
### Feature and motivation

If I add 

```
    dns:
      - "x.x.x.xxx"
    dns_search: testdomain.com
```

in my docker-compose file, the node-docker container's resol.conf file is updated and the container is able to resolve names using machine shortnames, and FQDN's from the custom DNS server, but any containers that are dynamically created are using the standard resolv.conf file and thus are forced to use IP addresses. 

It would be great if the child containers could copy this file, or, an option was provided to allow further customization of the child containers that the node-docker container creates. 

### Usage example

Would allow usage of dynamic grid and the use of domain names on a custom DNS server, ie local intranet, rather than being forced to use IP addresses

### LINKED ISSUE #2143: [🐛 Bug]: I can't mount the volume in dynamic gird for browser node
### What happened?

I am setting up an environment for selenium grid. I want to use selenium dynamic grid.

My test case is loading an attachment on a web page.

I tried to attach the volume in docker compose, but the attachments are only visible in node docker. Not in dynamically created browser containers.

I also tried using the configuration in config.toml and it still doesn't work:

[node]
max-sessions = 2
[docker]
configs = [
    "selenium/standalone-chrome:4.18.0", "{\"browserName\": \"chrome\"}"
    ]
url = "http://127.0.0.1:2375"
video-image = "selenium/video:ffmpeg-6.1-20231219"
volumes = ["path_on_disk:/opt/selenium/attachments"]

### Command used to start Selenium Grid with Docker (or Kubernetes)

```shell
version: "3"
services:
  node-docker:
    image: selenium/node-docker:4.18.0
    volumes:
      - ./assets:/opt/selenium/assets
      - ./attachments:/opt/selenium/attachments
      - ./config.toml:/opt/bin/config.toml
      - /var/run/docker.sock:/var/run/docker.sock
    depends_on:
      - selenium-hub
    environment:
      - SE_EVENT_BUS_HOST=selenium-hub
      - SE_EVENT_BUS_PUBLISH_PORT=4442
      - SE_EVENT_BUS_SUBSCRIBE_PORT=4443
      - SE_VNC_NO_PASSWORD=1

  selenium-hub:
    image: selenium/hub:4.18.0
    container_name: selenium-hub
    ports:
      - "4442:4442"
      - "4443:4443"
      - "4444:4444"
```

### Relevant log output

```shell
.
```

### Operating System

Windows 11

### Docker Selenium version (image tag)

4.18.0

### Selenium Grid chart version (chart version)

_No response_

### PR DESCRIPTION
**Pull request by VietND96**
## **User description**
**Thanks for contributing to Selenium!**
**A PR well described will help maintainers to quickly review and merge it**

Before submitting your PR, please check our [contributing](https://github.com/SeleniumHQ/selenium/blob/trunk/CONTRIBUTING.md) guidelines.
Avoid large PRs, help reviewers by making them as simple and short as possible.

### Motivation and Context
Fix: https://github.com/SeleniumHQ/docker-selenium/issues/2143
Fix: https://github.com/SeleniumHQ/docker-selenium/issues/2098
Fix: https://github.com/SeleniumHQ/docker-selenium/issues/1812
Fix: https://github.com/SeleniumHQ/docker-selenium/issues/1642

Docker compose file to deploy node-docker looks like

```yaml
services:
  node-docker:
    image: ${NAMESPACE}/node-docker:${TAG}
    user: root
    volumes:
      - ./videos:/opt/selenium/assets
      - ./videos/config.toml:/opt/bin/config.toml
      - /var/run/docker.sock:/var/run/docker.sock
      - ./videos/Downloads:/home/seluser/Downloads
    dns:
      - 8.8.8.8
      - 8.8.4.4
    dns_search: selenium-grid.local
    extra_hosts:
      - "prod.domain.com:${HOST_IP}"
    depends_on:
      - selenium-hub
    environment:
      - SE_EVENT_BUS_HOST=selenium-hub
      - SE_EVENT_BUS_PUBLISH_PORT=4442
      - SE_EVENT_BUS_SUBSCRIBE_PORT=4443
      - SE_NODE_ENABLE_MANAGED_DOWNLOADS=true
      - SE_LOG_LEVEL=${LOG_LEVEL}
```
`config.toml` looks like
```toml
[docker]
configs = [
    "${NAMESPACE}/standalone-firefox:${TAG}", '{"browserName": "firefox", "platformName": "linux"}',
    "${NAMESPACE}/standalone-chrome:${TAG}", '{"browserName": "chrome", "platformName": "linux"}',
    "${NAMESPACE}/standalone-edge:${TAG}", '{"browserName": "MicrosoftEdge", "platformName": "linux"}'
    ]

host-config-keys = ["Dns", "DnsOptions", "DnsSearch", "ExtraHosts", "Binds"]

url = "http://127.0.0.1:2375"
video-image = "${NAMESPACE}/video:${VIDEO_TAG}"

[node]
enable-managed-downloads = true
override-max-sessions = true
max-sessions = 4
```

With the config key `host-config-keys` under section `[docker]` in a `config.toml` file (or CLI option `--docker-host-config-keys`). Users can specify a list of docker host configuration keys that should be passed to browser containers.

Valid key names for Docker host config can be found in the Docker API [documentation](https://docs.docker.com/engine/api/latest/#tag/Container/operation/ContainerCreate) or via the command `docker inspect` the node-docker container.

Steps to debug

Set the log-level to `FINE`, and you are able to see the `Container config` looks like

> 10:15:38.988 DEBUG [DockerSessionFactory.createBrowserContainer] - Container config: ContainerConfig{image=Image{summary=org.openqa.selenium.docker.internal.ImageSummary@3bcbb589}, portBindings={}, envVars={SE_EVENT_BUS_PUBLISH_PORT=4442, SE_EVENT_BUS_HOST=selenium-hub, SE_REJECT_UNSUPPORTED_CAPS=false, SE_EVENT_BUS_SUBSCRIBE_PORT=4443, TZ=UTC, SE_NODE_ENABLE_MANAGED_DOWNLOADS=false, SE_OTEL_TRACES_EXPORTER=otlp, SE_BIND_HOST=false, SE_OTEL_JAVA_GLOBAL_AUTOCONFIGURE_ENABLED=true, SE_OTEL_SERVICE_NAME=selenium-node-docker, SE_LOG_LEVEL=FINE}, volumeBinds={}, networkName=tests_default, devices=[], autoRemove=true, shmSize=2147483648, hostConfig={ExtraHosts=[prod.domain.com:10.1.0.178], Binds=[/home/runner/work/docker-selenium/docker-selenium/tests/videos:/opt/selenium/assets:rw, /home/runner/work/docker-selenium/docker-selenium/tests/videos/config.toml:/opt/bin/config.toml:rw, /var/run/docker.sock:/var/run/docker.sock:rw, /home/runner/work/docker-selenium/docker-selenium/tests/videos/Downloads:/home/seluser/Downloads:rw], Dns=[8.8.8.8, 8.8.4.4], DnsSearch=[selenium-grid.local]}}

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### Checklist

- [x] I have read the [contributing](https://github.com/SeleniumHQ/selenium/blob/trunk/CONTRIBUTING.md) document.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

___

## **Type**
enhancement

___

## **Description**
- Added support for Docker host configurations in Selenium Grid's Docker integration, allowing for more granular control over browser containers.
- Introduced `hostConfig` in `ContainerConfig` and `ContainerInfo` to manage and apply Docker container configurations.
- Added a new flag `--docker-host-config-keys` to specify which host configuration keys should be passed to browser containers.
- Enhanced `DockerSessionFactory` to apply specified host configurations to browser containers, improving flexibility in container management.

___

## **Changes walkthrough**
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement
</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>ContainerConfig.java</strong><dd><code>Enhance ContainerConfig with Host Configuration Support</code>&nbsp; &nbsp; </dd></summary>
<hr>
      
java/src/org/openqa/selenium/docker/ContainerConfig.java

<li>Added support for <code>hostConfig</code> in <code>ContainerConfig</code> to manage Docker <br>container configurations.<br> <li> Introduced a new constructor and a method <code>getHostConfig</code> to filter and <br>apply host configurations based on provided keys.<br>

</details>
    

  </td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/13804/files#diff-ef3399f1ef34da7c97a287ae43ca7feda14f6d23189602a4a1cf70acb8709040">+34/-0</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>ContainerInfo.java</strong><dd><code>Include Host Config in ContainerInfo</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
java/src/org/openqa/selenium/docker/ContainerInfo.java

<li>Added <code>hostConfig</code> map to store Docker host configurations.<br> <li> Updated constructor to include <code>hostConfig</code> and added a getter method <br>for it.<br>

</details>
    

  </td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/13804/files#diff-121b8a2a0142c3abe8c3549cd7a4da52a1e88f422157efb0bbefe64b1cf75cb7">+11/-1</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>InspectContainer.java</strong><dd><code>Update InspectContainer to Fetch HostConfig</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
java/src/org/openqa/selenium/docker/v1_41/InspectContainer.java

<li>Modified <code>apply</code> method to include <code>hostConfig</code> in <code>ContainerInfo</code> <br>instantiation.<br>

</details>
    

  </td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/13804/files#diff-23f17ebe7cc55d5195846a286962a28abb1ac0a830dca39a149f824d706ee8cc">+2/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>DockerFlags.java</strong><dd><code>Add Docker Host Config Keys Flag</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
java/src/org/openqa/selenium/grid/node/docker/DockerFlags.java

<li>Introduced <code>--docker-host-config-keys</code> flag to specify Docker container <br>host configuration keys.<br>

</details>
    

  </td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/13804/files#diff-73c455146a4c6b8003174819c76874c0ea3f8d8e03874d447a4612145cdffb5e">+12/-0</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>DockerOptions.java</strong><dd><code>Implement Host Config Keys Handling in DockerOptions</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
java/src/org/openqa/selenium/grid/node/docker/DockerOptions.java

<li>Added handling for <code>hostConfigKeys</code> to filter and apply specific Docker <br>host configurations.<br>

</details>
    

  </td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/13804/files#diff-5a4e578c5c5711a9b26d92a059bfe58b426177f8d2fc68aaaa79cd0fbef456bd">+12/-1</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>DockerSessionFactory.java</strong><dd><code>Enhance DockerSessionFactory with Host Config Management</code>&nbsp; </dd></summary>
<hr>
      
java/src/org/openqa/selenium/grid/node/docker/DockerSessionFactory.java

<li>Added <code>hostConfig</code> and <code>hostConfigKeys</code> to manage Docker host <br>configurations for browser containers.<br> <li> Updated <code>createBrowserContainer</code> method to apply host configurations <br>based on the provided keys.<br>

</details>
    

  </td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/13804/files#diff-a05d31304c9c75c40d9d11ef388a2c669f1f8c35e943ed8dcbc32e1da7694c5e">+14/-4</a>&nbsp; &nbsp; </td>

</tr>                    
</table></td></tr></tr></tbody></table>

___

> ✨ **PR-Agent usage**:
>Comment `/help` on the PR to get a list of all available PR-Agent tools and their descriptions

### CODE REVIEW

**Comment by VietND96** [comment]
@diemol, could you please review this?

File: java/src/org/openqa/selenium/docker/ContainerConfig.java [hunk 1]
```diff
@@ -59,6 +81,7 @@ public ContainerConfig(
     this.networkName = networkName;
     this.autoRemove = true;
     this.shmSize = shmSize;
+    this.hostConfig = hostConfig != null ? hostConfig : ImmutableMap.of();
```
Comment (diemol) [inline java/src/org/openqa/selenium/docker/ContainerConfig.java]: Is there a chance `hostConfig` is null? Why don't we do a `Require.nonNull`?

File: java/src/org/openqa/selenium/grid/node/docker/DockerFlags.java [hunk 2]
```diff
@@ -69,6 +69,18 @@ public class DockerFlags implements HasRoles {
...
+      description =
+          "Node container host config keys to be passed to the browsers container. Keys name can be"
+              + " found in the Docker API documentation, or by running `docker inspect` the"
+              + " node-docker container.")
+  @ConfigValue(
+      section = DockerOptions.DOCKER_SECTION,
+      name = "host-config-keys",
+      example = "[\"Dns\", \"DnsOptions\", \"DnsSearch\", \"ExtraHosts\"]")
+  private List<String> hostConfigKeys;
+
```
Comment (diemol) [inline java/src/org/openqa/selenium/grid/node/docker/DockerFlags.java]: We need to add this to the docs.

**Review by diemol — COMMENTED** [review]
Thank you, @VietND96!

Just a comment in one of the changes.

File: java/src/org/openqa/selenium/grid/node/docker/DockerFlags.java (hunk 2 above)
Comment (VietND96) [reply java/src/org/openqa/selenium/grid/node/docker/DockerFlags.java]: Sure, I will add this to docs repo and also docker-selenium.

File: java/src/org/openqa/selenium/docker/ContainerConfig.java (hunk 1 above)
Comment (VietND96) [reply java/src/org/openqa/selenium/docker/ContainerConfig.java]: Yes, there is no chance `hostConfig` is null
Require.nonNull is also checked in ContainerConfig `this.hostConfig = Require.nonNull("Host config", hostConfig);`

**Review by diemol — APPROVED** [review]
Thank you, @VietND96!

**Pull request merged**